styx-scheduler-service/src/main/java/com/spotify/styx/state/QueuedStateManager.java

```diff
@@ -209,6 +225,22 @@ private void initialize(WorkflowInstance workflowInstance) {
     }
   }
 
+  private void verifyCounter(Event event, long expectedCounter, RunState currentRunState) {
+    final long currentCounter = currentRunState.counter();
+    if (currentCounter > expectedCounter) {
+      final String message = "Stale event encountered. Expected counter is "
+                             + expectedCounter  + " but current counter is "
+                             + currentCounter + ". Discarding event " + event;
+      LOG.warn(message);
+      throw new StaleEventException(message);
+    } else if (currentCounter < expectedCounter) {
```

Comment:
Do we really expect this to happen?

Reply:
This shouldn't happen, but I think it's good to cover all possibilities and log accordingly

Reply:
Then let’s at least have a comment like “this should never happen” so that later on when we read the code we won’t spend too much brain cycle trying to understand why.